## Using targeted sequence capture to identify evolutionary relationships among poplars and willows

This repository contains the scripts and analysis notebooks used for the
creation of a novel genotyping platform that I designed, and its
implementation reported in my paper
**Targeted sequence capture array for phylogenetics and population genomics in the Salicaceae**, published in the journal [Applications in Plant Sciences](https://doi.org/10.1002/aps3.11394) in 2020.

### Contents

1. [README.md](https://github.com/BrianSanderson/phylo-seq-cap/blob/master/README.md): this file

2. [notebooks/](https://github.com/BrianSanderson/phylo-seq-cap/blob/master/notebooks/): Notebooks that detail the analyses reported in the manuscript in Jupyter Notebook format

3. [probes/](https://github.com/BrianSanderson/phylo-seq-cap/blob/master/probes/): The sequence of probes used for target capture, as well as a target reference file used by HybPiper for gene sequence assembly

4. [scripts/](https://github.com/BrianSanderson/phylo-seq-cap/blob/master/scripts): Python scripts I wrote to design the sequence capture array, and to analyze the resulting data

The raw read data obtained through targeted sequence capture will be available at the [NCBI short read archive](https://www.ncbi.nlm.nih.gov/sra/PRJNA627181) after the manuscript has been accepted for publication.
